x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/amazonbedrock/AmazonBedrockInferenceClientCache.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.external.amazonbedrock;
+
+import com.amazonaws.http.IdleConnectionReaper;
+
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xpack.inference.services.amazonbedrock.AmazonBedrockModel;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+import static java.util.Collections.emptyMap;
+
+public final class AmazonBedrockInferenceClientCache implements AmazonBedrockClientCache {
+
+    private final BiFunction<AmazonBedrockModel, TimeValue, AmazonBedrockBaseClient> creator;
+    private volatile Map<Integer, AmazonBedrockBaseClient> clientsCache = emptyMap();
```

Comment:
What are your thoughts on using a `ConcurrentHashMap` here? I think if we use that we might be able to void most of the locking by using the `compute*` or `put*` style methods. If we went with a `ConcurrentHashMap` we lose the immutability I think though.

`flushExpiredClients` would be called on multiple threads so we'd probably still need locking around that 🤔 

I suppose another option would be to have a dedicated thread handle cleaning up the expired clients.

Reply:
I'm good with `ConcurrentHashMap` - I think that's a totally safer option to use here... not too worried about the immutability as the only place it would be affected is in the `releaseCachedClients()` call with the `clientsCache = emptyMap();` - but, that's easily changed to `clientsCache.clear()`